serverless/src/createProposal/__tests__/handler.test.js

```diff
@@ -0,0 +1,185 @@
+import { mockClient } from 'aws-sdk-client-mock'
+import { PutObjectCommand, S3Client } from '@aws-sdk/client-s3'
+
+import createProposal from '../handler'
+
+const s3ClientMock = mockClient(S3Client)
+
+let consoleErrorSpy
+
+beforeEach(() => {
+  vi.clearAllMocks()
+  s3ClientMock.reset()
+  consoleErrorSpy = vi.spyOn(console, 'error').mockImplementation(() => {})
+})
+
+afterEach(() => {
+  consoleErrorSpy.mockRestore()
+})
+
+describe('createProposal', () => {
+  const validProposal = {
+    id: 'test-proposal-id',
+    providerId: 'test-provider-id',
+    shortName: 'Test Proposal',
+    entryTitle: 'Test Proposal Title',
+    proposalStatus: 'DRAFT',
+    requestType: 'CREATE',
+    submitterId: 'test-submitter-id',
+    createdAt: '2023-05-20T12:00:00Z',
+    updatedAt: '2023-05-20T12:00:00Z',
+    draft: {
+      lastUpdated: '2023-05-20T12:00:00Z',
+      comment: 'Initial draft'
+    }
+  }
+  test('When given a valid proposal, should save it to S3 and return success', async () => {
+    s3ClientMock.on(PutObjectCommand).resolves({
+      $metadata: {
+        httpStatusCode: 200,
+        requestId: undefined,
+        extendedRequestId: undefined,
+        cfId: undefined,
+        attempts: 1,
+        totalRetryDelay: 0
+      },
+      ETag: '"1a7e08244b933e4fea1f920da4988500"'
+    })
+
+    const event = {
+      body: JSON.stringify(validProposal)
+    }
+
+    const response = await createProposal(event)
+
+    expect(response.statusCode).toBe(200)
+    expect(JSON.parse(response.body)).toEqual({ id: 'test-proposal-id' })
+  })
+
+  test('When S3 client is not initialized, should create a new S3 client', async () => {
+    s3ClientMock.on(PutObjectCommand).resolves({
+      $metadata: { httpStatusCode: 200 }
+    })
+
+    const event = {
+      body: JSON.stringify(validProposal)
+    }
+
+    await createProposal(event)
```

Comment:
Why do we need to do this twice here? and not just call it once then do expect(s3ClientMock.calls()).toHaveLength(1)

Reply:
Done, removed 2. call